broker/query_compiler.go

```diff
@@ -0,0 +1,123 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package broker
+
+import (
+	"fmt"
+	common2 "github.com/uber/aresdb/metastore/common"
+	"github.com/uber/aresdb/query/common"
+	"github.com/uber/aresdb/query/expr"
+	"github.com/uber/aresdb/query/sql"
+	"github.com/uber/aresdb/utils"
+	"net/http"
+)
+
+const (
+	nonAggregationQueryLimit = 1000
+)
+
+// QueryContext is broker query context
+type QueryContext struct {
+	RawQuery              string
+	AQLQuery              *common.AQLQuery
+	IsNonAggregationQuery bool
+	Writer                http.ResponseWriter
+	Error                 error
+}
+
+// NewQueryContext creates new query context
+func NewQueryContext(sqlQuery string, w http.ResponseWriter) *QueryContext {
+	ctx := QueryContext{
+		RawQuery: sqlQuery,
+		Writer:   w,
+	}
+	return &ctx
+}
+
+// Compile sql to AQL and extract information for routing
+func (c *QueryContext) Compile(schemaMutator common2.TableSchemaReader) {
+	c.AQLQuery, c.Error = sql.Parse(c.RawQuery, utils.GetLogger())
+	if c.Error != nil {
+		return
+	}
+
+	// validate main table
+	var err error
+	mainTableName := c.AQLQuery.Table
+	_, err = schemaMutator.GetTable(mainTableName)
+	if err != nil {
+		c.Error = utils.StackError(err, fmt.Sprintf("err finding main table %s", mainTableName))
+		return
+	}
+	// validate foreign table names
+	for _, join := range c.AQLQuery.Joins {
+		_, err = schemaMutator.GetTable(join.Table)
+		if err != nil {
+			c.Error = utils.StackError(err, fmt.Sprintf("err finding join table %s", join.Table))
+			return
+		}
+	}
+
+	c.proessMeasures()
+
+	return
+}
+
+func (c *QueryContext) proessMeasures() {
+	var err error
+
+	// Measures.
+	for i, measure := range c.AQLQuery.Measures {
+		measure.ExprParsed, err = expr.ParseExpr(measure.Expr)
+		if err != nil {
+			c.Error = utils.StackError(err, "Failed to parse measure: %s", measure.Expr)
+			return
+		}
+		c.AQLQuery.Measures[i] = measure
+	}
+	utils.GetLogger().Debug(c.AQLQuery.Measures[0].ExprParsed)
+
+	// ony support 1 measure for now
+	if len(c.AQLQuery.Measures) != 1 {
+		c.Error = utils.StackError(nil, "expect one measure per query, but got %d",
+			len(c.AQLQuery.Measures))
+		return
+	}
+
+	if _, ok := c.AQLQuery.Measures[0].ExprParsed.(*expr.NumberLiteral); ok {
+		c.IsNonAggregationQuery = true
+		// in case user forgot to provide limit
+		if c.AQLQuery.Limit == 0 {
+			c.AQLQuery.Limit = nonAggregationQueryLimit
+		}
+		return
+	}
+
+	// TODO: handle limit properly
```

Comment:
wrong comment position?

Reply:
not really.. not sure what's the best place to put this TODO, as it may require chages in multiple places. just put somewhere so I will remember.